src/Oscoin/Protocol.hs

```diff
@@ -0,0 +1,276 @@
+-- the input action and return the final result.
+withProtocol
+    :: forall c tx s a m.
+       Orphanage c tx s
+    -> Validate c tx s
+    -> (Block c tx (Sealed c s) -> Score)
+    -> BlockStore c tx s m
+    -> Consensus.Config
+    -> (Protocol c tx s m -> m a)
+    -> m a
+withProtocol o validateFull scoreBlock bs config action =
+    let protocol = Protocol bs o validateFull scoreBlock config
+    in action protocol
+
+-- | Initialises and runs the 'Protocol', by spinning an asynchronous worker
+-- which executes the 'fetchNextBlock' function to pull new blocks from outside
+-- and steps the protocol.
+runOscoinProtocol
+    :: forall c tx s a.
+       ( Ord (Hash c)
+       , Ord s
+       , HasHashing c
+       , Buildable (Hash c)
+       )
+    => Validate c tx s
+    -> (Block c tx (Sealed c s) -> Score)
+    -- ^ A function to score a block.
+    -> Telemetry.Handle
+    -> BlockStore c tx s IO
+    -> Consensus.Config
+    -> (Handle c tx s IO -> IO a)
+    -> IO a
+runOscoinProtocol validateFull scoreBlock telemetry bs config use =
+  bracket acquire dispose (\(hdl, _worker) -> use hdl)
+  where
+    acquire :: IO (Handle c tx s IO, Async ())
+    acquire = do
+        let o = O.emptyOrphanage scoreBlock
+        proto               <- newMVar (Protocol bs o validateFull scoreBlock config)
+        incomingBlocksQueue <- atomically $ newTBQueue 64
+        let hdl = Handle
+              { dispatchBlockSync = \blk -> do
+                  events <- modifyMVar proto $ \p -> stepProtocol p blk
+                  forM_ events (Telemetry.emit telemetry)
+              , dispatchBlockAsync = atomically . writeTBQueue incomingBlocksQueue
+              }
+
+        worker <- Async.async $ do
+                      block  <- atomically $ readTBQueue incomingBlocksQueue
+                      events <- modifyMVar proto $ \p -> stepProtocol p block
+                      forM_ events (Telemetry.emit telemetry)
+        pure (hdl, worker)
+
+    dispose :: (Handle c tx s IO, Async ()) -> IO ()
+    dispose (_, worker) =
+        -- TODO(adn) Log/emit that we are stopping the protocol runner.
+        Async.cancel worker
+
+-- | Main powerhouse of the 'Protocol' module. In order, this endless
+-- loop does the following:
+--
+-- 1. Gets incoming blocks from the outside;
+-- 2. Performs full (extrinsic) block validation;
+-- 3. Stores validated blocks into the BlockStore or as orphans;
+-- 4. Performs chain selection.
+stepProtocol
+    :: forall c tx s m.
+       ( Ord (BlockHash c)
+       , Ord s
+       , HasHashing c
+       , Buildable (Hash c)
+       , Monad m
+       )
+    => Protocol c tx s m
+    -> Block c tx (Sealed c s)
+    -> m (Protocol c tx s m, [NotableEvent])
+stepProtocol mgr incomingBlock = do
+    -- Step 2. Try to store the block; if this is recognised as to be an
+    -- orphan, we store it immediately (/PRECONDITION/: the block has been
+    -- pre-validate at the gossip level). If this block is scheduled for
```

Comment:
I'd move the precondition to the haddock, or else it won't be as useful I think.

Reply:
It's also not actually a pre-condition if I understand correctly, ie. the function will not fail if the block wasn't basic-validated, or am I wrong? If so, I'd turn it around and add `Nb. Does not perform basic validation`